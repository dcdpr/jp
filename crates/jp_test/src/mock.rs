use std::{
    env, fs,
    path::{Path, PathBuf},
};

use httpmock::{MockServer, RecordingRuleBuilder};
use saphyr::{LoadableYamlNode, Yaml, YamlEmitter};

/// A recorder for HTTP requests.
pub struct Vcr {
    forward_to: String,
    fixtures: PathBuf,
    recording: bool,
}

impl Vcr {
    /// Create a new recorder.
    pub fn new(forward_to: impl Into<String>, fixtures: impl Into<PathBuf>) -> Self {
        Self {
            forward_to: forward_to.into(),
            fixtures: fixtures.into(),
            recording: env::var("RECORD").is_ok(),
        }
    }

    /// Enable recording for the current test.
    pub fn record(&mut self) {
        self.recording = true;
    }

    /// Enable playback for the current test.
    pub fn playback(&mut self) {
        self.recording = false;
    }

    /// Set recording mode for the current test.
    pub fn set_recording(&mut self, recording: bool) {
        self.recording = recording;
    }

    /// Record a test cassette.
    pub async fn cassette<R, F, Fut, O>(
        &self,
        name: &str,
        rule_builder: R,
        test_fn: F,
    ) -> Result<(), Box<dyn std::error::Error>>
    where
        R: FnOnce(RecordingRuleBuilder),
        F: FnOnce(bool, String) -> Fut,
        Fut: Future<Output = O>,
        O: std::fmt::Debug,
    {
        let fixture = self.fixtures.join(format!("{name}.yml"));
        let server = MockServer::start_async().await;

        if self.recording {
            server
                .forward_to_async(&self.forward_to, |rule| {
                    rule.filter(|when| {
                        when.any_request();
                    });
                })
                .await;

            let recording = server.record_async(rule_builder).await;

            let out = test_fn(true, server.base_url()).await;
            self.verify(name, out);

            let temp_path = recording.save_to_async(&self.fixtures, name).await?;

            if temp_path != fixture && temp_path.exists() {
                fs::rename(temp_path, &fixture)?;
            }

            modify_fixture(&fixture)?;
        } else if fixture.exists() {
            server.playback_async(&fixture).await;
            let out = test_fn(false, server.base_url()).await;
            self.verify(name, out);
        } else {
            return Err(format!("Recording not found at {}", fixture.display()).into());
        }

        Ok(())
    }

    fn verify<T: std::fmt::Debug>(&self, name: &str, expr: T) {
        insta::with_settings!({ snapshot_path => &self.fixtures, prepend_module_to_snapshot => false }, {
            insta::assert_debug_snapshot!(name, expr);
        });
    }
}

fn modify_fixture(fixture: &Path) -> Result<(), Box<dyn std::error::Error>> {
    let contents = fs::read_to_string(fixture)?;
    let mut document = Yaml::load_from_str(&contents)?;
    for yaml in &mut document {
        let Some(header) = yaml
            .as_mapping_get_mut("then")
            .and_then(|then| then.as_mapping_get_mut("header"))
            .and_then(|header| header.as_sequence_mut())
        else {
            return Ok(());
        };

        header.retain(|header| {
            header.as_mapping_get("name").and_then(|v| v.as_str()) == Some("content-type")
        });
    }

    let mut buf = String::new();
    for yaml in &document {
        YamlEmitter::new(&mut buf).dump(yaml)?;
        buf.push('\n');
    }
    fs::write(fixture, buf.trim_start_matches("---\n"))?;

    Ok(())
}
