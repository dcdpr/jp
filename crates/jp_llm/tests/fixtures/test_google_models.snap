---
source: crates/jp_test/src/mock.rs
expression: expr
---
Ok(
    [
        ModelDetails {
            provider: Google,
            slug: "embedding-gecko-001",
            context_window: Some(
                1024,
            ),
            max_output_tokens: Some(
                1,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-1.5-pro-latest",
            context_window: Some(
                2000000,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-1.5-pro-002",
            context_window: Some(
                2000000,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-1.5-pro",
            context_window: Some(
                2000000,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-1.5-flash-latest",
            context_window: Some(
                1000000,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-1.5-flash",
            context_window: Some(
                1000000,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-1.5-flash-002",
            context_window: Some(
                1000000,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-1.5-flash-8b",
            context_window: Some(
                1000000,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-1.5-flash-8b-001",
            context_window: Some(
                1000000,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-1.5-flash-8b-latest",
            context_window: Some(
                1000000,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-pro-preview-03-25",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 128,
                    max_tokens: Some(
                        32768,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash-preview-05-20",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash-lite-preview-06-17",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-pro-preview-05-06",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 128,
                    max_tokens: Some(
                        32768,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-pro-preview-06-05",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 128,
                    max_tokens: Some(
                        32768,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-pro",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 128,
                    max_tokens: Some(
                        32768,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash-exp",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash-001",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash-lite-001",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash-lite",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash-lite-preview-02-05",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash-lite-preview",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-pro-exp",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-pro-exp-02-05",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-exp-1206",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash-thinking-exp-01-21",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash-thinking-exp",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash-thinking-exp-1219",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash-preview-tts",
            context_window: Some(
                8192,
            ),
            max_output_tokens: Some(
                16384,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-pro-preview-tts",
            context_window: Some(
                8192,
            ),
            max_output_tokens: Some(
                16384,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 128,
                    max_tokens: Some(
                        32768,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "learnlm-2.0-flash-experimental",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                32768,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemma-3-1b-it",
            context_window: Some(
                32768,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemma-3-4b-it",
            context_window: Some(
                32768,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemma-3-12b-it",
            context_window: Some(
                32768,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemma-3-27b-it",
            context_window: Some(
                131072,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemma-3n-e4b-it",
            context_window: Some(
                8192,
            ),
            max_output_tokens: Some(
                2048,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemma-3n-e2b-it",
            context_window: Some(
                8192,
            ),
            max_output_tokens: Some(
                2048,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash-lite",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash-image-preview",
            context_window: Some(
                32768,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "embedding-001",
            context_window: Some(
                2048,
            ),
            max_output_tokens: Some(
                1,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "text-embedding-004",
            context_window: Some(
                2048,
            ),
            max_output_tokens: Some(
                1,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-embedding-exp-03-07",
            context_window: Some(
                8192,
            ),
            max_output_tokens: Some(
                1,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-embedding-exp",
            context_window: Some(
                8192,
            ),
            max_output_tokens: Some(
                1,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-embedding-001",
            context_window: Some(
                2048,
            ),
            max_output_tokens: Some(
                1,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "aqa",
            context_window: Some(
                7168,
            ),
            max_output_tokens: Some(
                1024,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "imagen-3.0-generate-002",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "imagen-4.0-generate-preview-06-06",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "imagen-4.0-ultra-generate-preview-06-06",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "imagen-4.0-generate-001",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "imagen-4.0-ultra-generate-001",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "imagen-4.0-fast-generate-001",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "veo-2.0-generate-001",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "veo-3.0-generate-preview",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "veo-3.0-fast-generate-preview",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "veo-3.0-generate-001",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "veo-3.0-fast-generate-001",
            context_window: Some(
                480,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash-preview-native-audio-dialog",
            context_window: Some(
                131072,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash-exp-native-audio-thinking-dialog",
            context_window: Some(
                131072,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.0-flash-live-001",
            context_window: Some(
                131072,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-live-2.5-flash-preview",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: None,
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash-live-preview",
            context_window: Some(
                1048576,
            ),
            max_output_tokens: Some(
                65536,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash-native-audio-latest",
            context_window: Some(
                131072,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
        ModelDetails {
            provider: Google,
            slug: "gemini-2.5-flash-native-audio-preview-09-2025",
            context_window: Some(
                131072,
            ),
            max_output_tokens: Some(
                8192,
            ),
            reasoning: Some(
                Supported {
                    min_tokens: 0,
                    max_tokens: Some(
                        24576,
                    ),
                },
            ),
            knowledge_cutoff: None,
        },
    ],
)
